jinahub/indexers/searcher/compound/FaissPostgresSearcher/faisspsql.py

```diff
@@ -2,50 +2,172 @@
 
 try:
-    from jinahub.indexers.storage.PostgreSQLStorage import (
-        PostgreSQLStorage,
-    )
-except:
-    from jina_executors.indexers.storage.PostgreSQLStorage import (
-        PostgreSQLStorage,
-    )
+    from jinahub.indexers.storage.PostgreSQLStorage import PostgreSQLStorage
+except ImportError:  # noqa: E722
+    from jina_executors.indexers.storage.PostgreSQLStorage import PostgreSQLStorage
 
 
 class FaissPostgresSearcher(Executor):
-    """A Compound Indexer made up of a FaissSearcher (for vectors) and a Postgres Indexer"""
+    """A Compound Indexer made up of a FaissSearcher (for vectors) and a Postgres
+    Indexer
+
+    :param dump_path: a path to a dump folder containing
+    the dump data obtained by calling jina_commons.dump_docs
+    :param startup_sync: whether to use the dump
+     function of PostgreSQLStorage, when dump_path is not provided
+    :param startup_sync_args: the arguments to be passed to the self.sync call on
+    startup (if startup_sync)
+    """
 
     def __init__(
         self,
-        dump_path=None,
+        dump_path: Optional[str] = None,
+        startup_sync: bool = False,
+        startup_sync_args: Optional[None] = None,
         **kwargs,
     ):
         super().__init__(**kwargs)
-        # when constructed from rolling update the dump_path is passed via a runtime_arg
+        self.logger = get_logger(self)
+
+        self.total_shards = self.runtime_args.parallel
+        if self.total_shards is None:
+            self.logger.warning(
+                'total_shards is None, rolling update '
+                'via PSQL import will not be possible.'
+            )
+
+        # when constructed from rolling update
+        # args are passed via runtime_args
         dump_path = dump_path or kwargs.get('runtime_args').get('dump_path')
-        self.logger = JinaLogger(getattr(self.metas, 'name', self.__class__.__name__))
+        startup_sync = startup_sync or kwargs.get('runtime_args').get('startup_sync')
+
         self._kv_indexer = None
         self._vec_indexer = None
-        if dump_path:
-            self._vec_indexer = FaissSearcher(dump_path=dump_path, **kwargs)
-            self._kv_indexer = PostgreSQLStorage(**kwargs)
+        self._init_kwargs = kwargs
+
+        (
+            self._kv_indexer,
+            self._vec_indexer,
+            startup_sync,
+            startup_sync_args,
+        ) = self._init_executors(dump_path, kwargs, startup_sync, startup_sync_args)
+        if startup_sync:
+            self.sync(parameters=startup_sync_args)
+
+    def _init_executors(self, dump_path, kwargs, startup_sync, startup_sync_args):
+        kv_indexer = PostgreSQLStorage(**kwargs)
+
+        if startup_sync_args is None and startup_sync:
+            startup_sync_args = {}
+        # if the user passes args for syncing, they must've meant to sync as well
+        if startup_sync_args and (startup_sync is False or startup_sync is None):
+            startup_sync = True
```

Comment:
This code smells. It cries, that we should only have `startup_sync_args` and not `startup_sync`. And `startup_sync_args` is also ugly, since it is not clear what to provide. Let's have a session tomorrow to see, how we can make this more user-friendly. I have a design in mind. We will do this in a follow-up PR.

Reply:
Much nicer :+1: